File: code/Java/interpreter/expr/BinaryIntExpr.java
package interpreter.expr;

public class BinaryIntExpr extends IntExpr {

    private IntExpr left;
    private IntOp op;
    private IntExpr right;

    public BinaryIntExpr(int line, IntExpr left, IntOp op, IntExpr right) {
        super(line);
        this.left = left;
        this.op = op;
        this.right = right;
    }

    public int expr() {
        int v1 = left.expr();
        int v2 = right.expr();

        switch (op) {
            case Add:
                return v1 + v2;
            case Sub:
                return v1 - v2;
            case Mul:
                return v1 * v2;
            case Div:
                return v1 / v2;
            case Mod:
            default:
                return v1 % v2;
        }
    }
    
}

File: code/C++/interpreter/expr/BinaryIntExpr.cpp
#include "BinaryIntExpr.h"

BinaryIntExpr::BinaryIntExpr(int line, IntExpr* left, enum BinaryIntExpr::Op op, IntExpr* right)
	: IntExpr(line), m_left(left), m_op(op), m_right(right) {
}

BinaryIntExpr::~BinaryIntExpr() {
	delete m_left;
	delete m_right;
}

int BinaryIntExpr::expr() {
	int v1 = m_left->expr();
	int v2 = m_right->expr();

	switch (m_op) {
		case BinaryIntExpr::ADD:
			return v1 + v2;
		case BinaryIntExpr::SUB:
			return v1 - v2;
		case BinaryIntExpr::MUL:
			return v1 * v2;
		case BinaryIntExpr::DIV:
			return v1 / v2;
		case BinaryIntExpr::MOD:
		default:
			return v1 % v2;
	}
}
